### DISCUSSION TITLE
Get precise SVG dimensions

### DISCUSSION THREAD

**Discussion by teohhanhui** [libvips/libvips#2022] [Ideas]
Is it possible to use [rsvg_handle_get_geometry_sub](https://developer.gnome.org/rsvg/unstable/rsvg-RsvgHandle.html#rsvg-handle-get-geometry-sub) instead of [rsvg_handle_get_dimensions](https://developer.gnome.org/rsvg/unstable/rsvg-RsvgHandle.html#rsvg-handle-get-dimensions)?

See https://gitlab.gnome.org/GNOME/librsvg/issues/250#note_367736

Related issue in librsvg: https://gitlab.gnome.org/GNOME/librsvg/issues/134

**Comment by teohhanhui** [comment]
But it seems like `vips_image_init_fields` expects `int` for the width and height :disappointed:

**Comment by jcupitt** [comment]
Hi @teohhanhui, 

That's right, libvips is a raster library and works in whole pixels. Sorry :( 

libvips lets you set arbitrary metadata, so we could add something like this to the svg loader:

```C
vips_image_set_double(output_image, "svg-width", 3.1415927);
```

Could that be a solution? sharp users could then do `image.get("svg-width")` to get this value back (I think).

**Comment by jcupitt** [comment]
... very happy to accept a PR for this feature, I should have said.

**Comment by teohhanhui** [comment]
> libvips is a raster library and works in whole pixels

But the calculation for the scaling would still benefit from using `double`, no?

**Comment by jcupitt** [comment]
Oh huh yes I guess that's true. I'll have a look tomorrow.

**Comment by jcupitt** [comment]
I had a quick look and this is still very bleeding edge. `rsvg_handle_get_geometry_sub()` is only in git master librsvg, and that will only build with rustc 1.31, which is too new for Ubuntu 18.10.

We should probably wait for this to be released and stabilize before adding support. 

I made a branch to experiment with this feature, if you'd like to make a PR.

**Comment by teohhanhui** [comment]
I believe it's been there since v2.45.1, which builds with rustc 1.28

https://gitlab.gnome.org/GNOME/librsvg/blob/2.45.1/librsvg/rsvg-handle.c#L1283

**Comment by jcupitt** [comment]
Ah you're right, I got 2.45 and 2.46 mixed up.

I updated that branch to fetch SVG dimensions with rsvg_handle_get_geometry_sub(), if possible. 

I don't have an SVG with fractional size, do you have an example file?

I didn't add the thing to attach the fractional size as extra metadata.

**Comment by teohhanhui** [comment]
Here you go: https://commons.wikimedia.org/wiki/File:REMAX_logo.svg

**Comment by jcupitt** [comment]
Thanks! 

I still get the same size from all these calls though:

https://github.com/libvips/libvips/commit/fa8c244b2511d4e0f94d23a2ce0c1b6d81b217e7

```
john@yingna ~/pics $ vipsheader REMAX_logo.svg 
dimensions.width = 1290
dimensions.height = 236
ink_rect.width = 1290
ink_rect.height = 236
logical_rect.width = 1290
logical_rect.height = 236
REMAX_logo.svg: 1290x236 uchar, 4 bands, srgb, svgload
```

That's with librsvg 2.45.5.
